### DISCUSSION TITLE
What's the best way to update with a dynamic set of columns?

### DISCUSSION THREAD

**Discussion by Yoric** [diesel-rs/diesel#3417] [Q&A] [answered]
In my current app, I have a table, say

```sql
CREATE TABLE FOO(
  id BIGSERIAL PRIMARY KEY,
  a TEXT NOT NULL,
  b BIGINT NOT NULL,
  c TEXT,
 -- and a few more columns
);
```

Now, in some function calls, I receive a

```rust
struct Diff {
   a: Option<String>,
   b: Option<i64>,
   c: Option<Option<String>>,
   // ...
}
```

What's the best way to apply a `Diff` to my table? 

```rust
let mut update = diesel::update(...).into_boxed();
if let Some(a) = diff.a {
  update = update.set(schema::foo::dsl::foo::a.eq(a)).into_boxed();
}
if let Some(b) = diff.b {
  update = update.set(schema::foo::dsl::foo::b.eq(b)).into_boxed();
}
// ...
```

or is there a more streamlined way to do that?

**Comment by weiznich** [comment, marked as answer]
Please checkout the ["All about updates" guide](https://diesel.rs/guides/all-about-updates.html) on our webpage. Especially the section about the `AsChangeset` derive.

**Reply by Yoric** [reply]
Thanks, exactly what I was looking for!

**Reply by Yoric** [reply]
@weiznich I don't see anything about updating many entries at once and `Vec<T>` doesn't seem to implement `AsChangeset` even if `T` does. Is there any way to update many values at once?

**Reply by weiznich** [reply]
You want to build a struct that implements `AsChangeset` and holds all possibly changed fields. Each field is an `Option<…>`. Then you initialize all fields to `None` + set them to `Some` one by one as required. Afterwards you pass everything to `diesel::update`, which will skip all `None` fields.

**Reply by Yoric** [reply]
So what you're saying is "call `diesel::update` many times, once per record I wish to update"?

**Reply by weiznich** [reply]
> So what you're saying is "call diesel::update many times, once per record I wish to update"?

No thats not what I wrote above nor is that whats described in the linked guide. Again: Please checkout how `Option` and `#[derive(AsChangeset)]` interact. That's described in detail in the linked guide and in the Api documentation. You likely do not even need to construct any other type than your input struct, just using `[#derive(AsChangesetl]` there should work.

**Reply by Yoric** [reply]
I suspect that I'm not being clear.

I have a hundred (for instance) records that need updating. I have a `struct FooChangeset` that derives `AsChangeset`. I have one hundred instances of `FooChangeset`, each of which contains all the changes to said record. These changes are independent from each other.

As far as I can tell, this scenario is not covered in the linked guide. Am I missing something?

**Reply by weiznich** [reply]
That was not clear from your question. Maybe consider always to include an more or less complete example of what you try to do.

Anyway: Batch updates are something that is not really supported by SQL. There are ways to make this somehow work, but there is not 1:1 mapping to SQL there. As of that it is a feature thats not implemented in diesel yet. See https://github.com/diesel-rs/diesel/discussions/2965 for additional discussion on this topic. Contributions there are also welcome.
